### ISSUE TITLE
constructors / destructors execute twice

### ISSUE THREAD

**Issue by joe-lawrence** [numactl/numactl#44]
(By reading the code, this looks like known behavior.  But I figured I could re-open and ask a few questions about it.)

While helping @sanskriti-s debug some of the valgrind complaints #43, we noticed strange behavior of the libnuma constructor and destructor routines.

If adding some printk debugging:
```
  diff --git a/libnuma.c b/libnuma.c
  index eb995678ad54..a80f851f65c6 100644
  --- a/libnuma.c
  +++ b/libnuma.c
  @@ -92,6 +92,7 @@ numa_init(void)
   {
          int max,i;
   
  +printf("DEBUG: %s\n", __func__);
          if (sizes_set)
                  return;
   
  @@ -111,6 +112,7 @@ numa_init(void)
   void __attribute__((destructor))
   numa_fini(void)
   {
  +printf("DEBUG: %s\n", __func__);
          FREE_AND_ZERO(numa_all_cpus_ptr);
          FREE_AND_ZERO(numa_possible_cpus_ptr);
          FREE_AND_ZERO(numa_all_nodes_ptr);
```
Things appear as expected when building a static libnuma.a:
```
  % make clean 
  % ./autogen.sh
  % ./configure --enable-static --enable-shared=no
  % make
  % ./numademo | grep DEBUG
  DEBUG: numa_init
  DEBUG: numa_fini
```
However, when building a dynamic shared libnuma.so, both are executed *twice*:
```
  % make clean 
  % ./autogen.sh
  % ./configure --enable-static=no --enable-shared=yes
  % make
  % LD_LIBRARY_PATH=$(pwd)/.libs .libs/numademo | grep DEBUG
  DEBUG: numa_init
  DEBUG: numa_init
  DEBUG: numa_fini
  DEBUG: numa_fini
```
Note that this 2x execution appears to be accounted for by commit ec6e4552 "Use constructors for numa_init/exit".

However, if we strip the explicit linker directive to mark the -init and -fini routines from the Makefile.am:
```
  diff --git a/Makefile.am b/Makefile.am
  index 1c4266d43f26..c16a50780002 100644
  --- a/Makefile.am
  +++ b/Makefile.am
  @@ -40,7 +40,7 @@ memhog_SOURCES = memhog.c util.c
   memhog_LDADD = libnuma.la
   
   libnuma_la_SOURCES = libnuma.c syscall.c distance.c affinity.c affinity.h sysfs.c sysfs.h rtnetlink.c rtnetlink.h versions.ldscript
  -libnuma_la_LDFLAGS = -version-info 1:0:0 -Wl,--version-script,$(srcdir)/versions.ldscript -Wl,-init,numa_init -Wl,-fini,numa_fini
  +libnuma_la_LDFLAGS = -version-info 1:0:0 -Wl,--version-script,$(srcdir)/versions.ldscript
   
   check_PROGRAMS = \
          test/distance 
```
both libnuma.a and libnuma.so test cases only execute the constructor/destructors *once*:
```
  % make clean 
  % ./autogen.sh
  % ./configure --enable-static --enable-shared=no
  % make
  % ./numademo | grep DEBUG
  DEBUG: numa_init
  DEBUG: numa_fini

  % ./autogen.sh
  % ./configure --enable-static=no --enable-shared=yes
  % make
  % LD_LIBRARY_PATH=$(pwd)/.libs .libs/numademo | grep DEBUG
  DEBUG: numa_init
  DEBUG: numa_fini
```
I did some testing with a minimal libmin.a and libmin.so to verify this behavior and I ended up with these results:

- libmin.so = init/fini always executed
- libmin.a  - main program doesn't reference any libmin code 
                = no init/fini
- libmin.a  - main program references libmin code 
                = init/fini called

We were curious as to what the intended libnuma behavior should be.  Can we safely remove the init/fini linker directives from Makefile.am?  We weren't sure, so we coded any #43 libnuma fixes with this double execution in mind.

-- Joe

**Comment by filbranden** [comment]
Is this fixed by #43? If so, can we close it?

**Comment by joe-lawrence** [comment]
Hi @filbranden - no, #43 included "libnuma: cleanup node cpu mask in destructor" which only added some additional destructor cleanup.  I opened this issue after trying to figure out why the elaborate `FREE_AND_ZERO` macros.  After digging through the git history and running some tests, my conclusion was that the additional linker directive in Makefile.am was causing multiple invocations of `numa_{init,fini}` when libnuma was built as a shared library.  `numa_fini` was coded to handle this situation by clearing pointers after freeing them...  so there really isn't a bug here, more of a head-scratching "was this really intentional?" question.
